src/blas/KokkosBlas1_dot.hpp

```diff
@@ -244,6 +244,7 @@ dot (const RV& R, const XMV& X, const YMV& Y,
   YMV_Internal Y_internal = Y;
 
   Impl::Dot<RV_Internal, XMV_Internal, YMV_Internal>::dot(R_internal, X_internal, Y_internal);
+  Kokkos::fence();
```

Comment:
Oh, I think I get it.  The problem is that if you use the version of `Kokkos::parallel_reduce` that takes the output as a View, rather than as a reference to host memory, then the reduction is asynchronous.

I would call this overkill, though.  Most dot products are for a single vector, in which case the version of `parallel_reduce` that takes a reference to host memory is the right overload to call.  If I call the overload of `dot` that takes an output View, then I, the caller, should be responsible for fencing.

FYI @trilinos/tpetra @kddevin

Reply:
I agree with mark here ..